docs/design/features/StringDeduplication.md

```diff
@@ -0,0 +1,196 @@
+
+- Reference equality
+
+**Runtime assistance to help with these scenarios**
+
+- [ ] Static analysis
+
+For .NET 5.0, there’s a static analysis [plan](https://github.com/dotnet/runtime/issues/30740). Verifying if string content is mutated can be part of that analysis suite. It would also be beneficial to have VS do this verification and give you a warning when you are mutating string content.
+
+To start with we will provide analysis for the following –
+
+1. Using the fixed keyword to modify string content.
+
+I’m seeing that there are almost 600 places in libraries that do “fixed (char*” but hopefully most of them do not actually modify the string content. We should definitely be encouraging folks to switch to using string.Create like what [PR#31700](https://github.com/dotnet/runtime/pull/31700) did.
+
+2. Using lock on a string object. 
+
+- [ ] Reference equality checks on strings
+
+Since `ReferenceEquals` is performance critical API, we cannot do checks in its implementation. Checks will be done by how it's jitted - when the JIT sees that there are two object references being compared for equality, when a special diagnostics flag is set, it can jit a call to a helper method to do that and that helper method can produce diagnostics when the code is trying to compare identity of two strings.
+
+- [ ] Additional checks in heap verification 
+
+Heap verification will now include checks to verify that no one changes the string content after it’s hash is computed. This can be turned on when a certain level of COMPlus_HeapVerify is specified.
+
+- [ ] Stress mode
+
+Instead of waiting till the productive moment to start the next deduping cycle, we can have a stress mode where we dedup randomly to catch problems sooner, same idea as GC stress to detect GC holes sooner. 
+
+We could even artificially create duplicates in this stress mode to find places that depend on object identity.
+
+- [ ] Backup nonconcurrent dedup mode
+
+If you are running into these problematic scenarios and cannot make changes, and you can tolerate longer STW (Stop-The-World) pauses, we will provide a mode where deduping is completely done in either a full blocking GC or during the STW pause of a BGC. Obviously this would make BGC’s STW pauses much, much longer. This would basically just call the function that the dedup thread calls, but in these STW pauses. If you use a CPU profiler you should be able to see preciously how long the dedup part takes during the STW pause.
+
+We will either provide this as an API (GC.DeduplicateStrings) or as a config that only does dedup during full blocking GCs. The former is flexible and can give you detailed info such as how much space deduplicating strings saved. You could call this leisurely at a time when you know longer latency can be tolerated (eg, the server instance is taken out of rotation so you can tolerate a long pause).
+
```

Comment:
Also mention potential future extension to deduplicate other types of objects? Described in https://github.com/dotnet/runtime/issues/12628

Reply:
sounds good